### DISCUSSION TITLE
v7 useFieldArray with optional field causes TypeScript issues

### DISCUSSION THREAD

**Discussion by LauSam09** [react-hook-form/react-hook-form#5301] [General] [answered]
So I scratched my head over this for a while before figuring out the problem!

In v6, these types were entirely valid:
```ts
interface Thing {
  name: string
}

interface Form {
  things?: Array<Thing>
}
```

However in v7, if the property being used for `useFieldArray` can be `undefined` then the types end up being `never`:
![Screenshot from 2021-05-23 18-16-29](https://user-images.githubusercontent.com/24778857/119270121-10e7b600-bbf3-11eb-890d-2c13498af3ff.png)

Not a big deal but might save some time for other people also making the migration :smiley:

**Comment by bluebill1049** [comment]
Thanks very much for sharing solution

**Reply by LauSam09** [reply]
No worries - I'll rename the discussion slightly to make it easier for people to find.

**Reply by bluebill1049** [reply]
If you have a codesandbox, post in a separate comment, we can mark yours as the answer.

**Comment by LauSam09** [comment, marked as answer]
See the two sandboxes below which demonstrate the issue - the only difference is one is using `v6` (and the old `ref` registration syntax) and the other `v7`:

https://codesandbox.io/s/usefieldarray-typescript-v6-8c95r
https://codesandbox.io/s/usefieldarray-typescript-v7-ku0t0

Making the change below in the `v7` sandbox resolves the issue.
```diff
interface Form {
- things?: Array<Thing>;
+ things: Array<Thing>;
}
```

**Reply by RanaOsamaAwadallah** [reply]
But what if we want it an unrequired field as it is?

**Reply by LauSam09** [reply]
> But what if we want it an unrequired field as it is?

@RanaOsamaAwadallah sorry for the delay. The best I can suggest is keeping the field on the interface as required, initialise it with an empty array, and then in your `onSubmit` handler, map an empty array back to `undefined`.

Does that solve your problem?
